website/blog/2025-07-07-segfault-dependency-update.mdx

```diff
@@ -0,0 +1,81 @@
+One of the culprits resulting in SEGFAULTs in the past was the conflict due to the usage of Apache Thrift and Facebook Thrift.
+However, this was not the root cause this time because we were able to reproduce this issue by just building the test without the Apache Thrift dependency installed.
+We were entering a new territory to investigate, and we were not sure where to start.
+
+We then compiled an example called *EchoService* in Facebook Thrift that was very similar to the *RemoteServer*, and it worked. Then we copied and compiled the Velox *RemoveServer*
+in Facebook Thrift and that worked too! So the culprit was likely in the compilation flags, which likely differed between Facebook Thrift and Velox.
+We enabled the verbose logging for both builds and were able to spot one difference. We saw the GCC *coroutines* flag being used in the Facebook Thrift build.
+
+We were also curious about the invocation of the destructor instead of the actual function. We wore our gdb gloves again and dumped the entire
+vtable for the RemoteServer class and its base classes. The vtables were different when it was built in Velox vs. Facebook Thrift.
+Specifically, the list of functions inherited from *ServiceHandlerBase* was different.
+
+The vtable for the *RemoteServer* handler in the Velox build had the following entries:
+
+```
+folly::SemiFuture<folly::Unit> semifuture_onStartServing()
+folly::SemiFuture<folly::Unit> semifuture_onStopRequested()
+Thunk ServiceHandlerBase::~ServiceHandlerBase
+Thunk ServiceHandlerBase::~ServiceHandlerBase
+```
+
+The vtable for the *RemoteServer* handler in the Facebook Thrift build had the following entries:
+
+```
+folly::coro::Task<void> co_onStartServing()
+folly::coro::Task<void> co_onStopRequested()
+folly::SemiFuture<folly::Unit> semifuture_onStartServing()
+folly::SemiFuture<folly::Unit> semifuture_onStopRequested()
+Thunk ServiceHandlerBase::~ServiceHandlerBase
+Thunk ServiceHandlerBase::~ServiceHandlerBase
+```
+
+Tying up both the pieces of evidence, we could conclude that Velox generated a different vtable structure compared to what Facebook Thrift (and thus ThriftServer) was built with.
+Looking around further, we noticed that the *ServiceHandlerBase* was conditionally adding functions based on the *-fcoroutines* flag.
+
+```
+class ServiceHandlerBase {
+  ....
+ public:
+#if FOLLY_HAS_COROUTINES
+  virtual folly::coro::Task<void> co_onStartServing() { co_return; }
+  virtual folly::coro::Task<void> co_onStopRequested() {
+    throw MethodNotImplemented();
+  }
+#endif
+
+  virtual folly::SemiFuture<folly::Unit> semifuture_onStartServing() {
+#if FOLLY_HAS_COROUTINES
+    return co_onStartServing().semi();
+#else
+    return folly::makeSemiFuture();
+#endif
+  }
+```
+
+As a result, the ThriftServer would access an incorrect function, the destructor instead of the expected initialization function, because the functions at the offset used were different, thus resulting in a SEGFAULT.
```

Comment:
We have both vtables above, explicitly calling out the offsets would probably make it easier to follow